---
title: "Dataset_Comparison_CCLE_NCI"
date: "`r Sys.Date()`"
output:
    html_document:
      code_folding: hide
      toc: true
      toc_depth: 2
      highlight: tango


---

# Setup

```{r setup,  echo=FALSE, cache=FALSE, warning= FALSE, message=FALSE}
pacman::p_load(knitr,cgwtools,tidyverse, MatrixCorrelation, here, ComplexHeatmap, circlize, 
               clusterProfiler,AnnotationDbi,org.Hs.eg.db,PCAtools, DelayedMatrixStats,png, gifski,
               gganimate)


## Global options
options(max.print="75")
opts_chunk$set(cache=FALSE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE)
opts_knit$set(root.dir = "./../")

```

# Overall Matrix agreement subsetted for cell lines ang genes in common and order. 
RPPA?


IDEAS
- PCA of eIF, eEF and heatmaps for all gens sorted by pathway, or clustered
  - lineage per heatmap
  - Heatmap CORUM
  - eIF per tissue type-Correlation with pathway expression?
  - Growth and hypoxia of each cell type both from RNAseq and Protein, 
    do they agree? are they releveant to Translational efficiency
  - Healthy samples Translational efficiency, Lineage characteristics in cancer?
  - Association plots: Cancer type specific marks not in healthy tissue.
  - Power Analysis with Subsetting
  

```{r Data_load_Subset}

if( #for intereactive running or for knitting
  getwd() == "C:/Users/skourtis/OneDrive - CRG - Centre de Regulacio Genomica/Bioinformatics Projects/multiOmics/Project_Codes"){
    Files_to_load <- (c("./../Project_Datasets/CCLE_OMICS.rda",
                "./../Project_Datasets/NCI60_OMICS.rda",
                "./../Project_Datasets/Metabolic_Pathway_annotations.RData"))
    HUMAN_9606_idmapping_file <- "./../Project_Datasets/HUMAN_9606_idmapping.dat"
    Derms_file = "./../Project_Datasets/Lineage_tissue_types.xlsx"
    Metabolite_mapping_file  = "./../Project_Datasets/CCLE_metabolites_mapped.csv"
} else {
  Files_to_load <- (c("./Project_Datasets/CCLE_OMICS.rda",
                      "./Project_Datasets/NCI60_OMICS.rda",
                      "./Project_Datasets/Metabolic_Pathway_annotations.RData"))
  HUMAN_9606_idmapping_file <- "./Project_Datasets/HUMAN_9606_idmapping.dat"
  Derms_file = "./Project_Datasets/Lineage_tissue_types.xlsx"
  Metabolite_mapping_file  = "./Project_Datasets/CCLE_metabolites_mapped.csv"
  }

if(intersect(map(Files_to_load, cgwtools::lsdata) %>% unlist(),
             ls()) %>% length() == 0){
  purrr::walk(Files_to_load, load,envir = .GlobalEnv)
} else {
  stop("Variable names in environment which clash with RData loading Names")
             }

#Mapping of Uniprots to all others sort of identifiers
HUMAN_9606_idmapping <- read_tsv(HUMAN_9606_idmapping_file,
                                 col_names = FALSE) %>%
    setNames(c("Uniprot", "Type", "ID"))


Derms <- openxlsx::read.xlsx(Derms_file, cols = c(1,6)) %>% na.omit()

## the file humankegg_df if currently used for subpathways instead of this
#Ordering SMPDB by pathway count and removing duplicates. 
#This allows proteins to be in the most abundance pathway that they are in
# SMPDB <- SMPDB_pathways %>% group_by(Pathway) %>%
#   summarise(count=n()) %>% 
#   right_join(SMPDB_pathways, by = 'Pathway') %>% 
#   arrange(-count) %>% 
#   subset(!duplicated(Gene_Name)) %>% .[,-c(2:4,6:9,11)] %>%
#   pivot_longer(-Pathway, names_to = "Type", values_to = "ID") %>%
#   setNames(c("Subpathway", "Type","ID"))




#subset hsa KEGG to Gene name
HUMAN_9606_idmapping_hsa <- HUMAN_9606_idmapping[str_detect(HUMAN_9606_idmapping$ID, "hsa:"),-2] %>% 
    left_join(HUMAN_9606_idmapping[HUMAN_9606_idmapping$Type == "Gene_Name", -2], by = "Uniprot") %>% 
    na.omit() %>% .[,-1] %>% setNames(c("KEGG_ID", "Gene.names"))

#pairwise correlation function used in main function
Correlation_pairwise <- function(x,y){
  cor(x,y,use = "pairwise", method = "pearson")
}

# Get lower triangle of the correlation matrix
  get_upper_tri<-function(cormat){
    cormat[lower.tri(cormat)] <- NA
    return(cormat)
  }
# Get upper triangle of the correlation matrix
  get_lower_tri<-function(cormat){
    cormat[upper.tri(cormat)] <- NA
    return(cormat)
  }
  
#compare 2 datasets with common Feature (rows) and Sample names (Columns)
### Change rerun value
Dataset_correlation <- function(Dataset_1,Dataset_2){
  # type = c("Transcriptomics",
  #                                                             "Proteomic",
  #                                                             "Metabolomics")
  #  type <- "Transcriptomics"
  # Type <- case_when(
  #     type == "Transcriptomics" ~ "Gene_name",
  #     type == "Proteomic" ~ "Uniprot",
  #     type == "Metabolomics" ~ "Metabolites",
  #     TRUE ~ as.character(type))

    #Dataset_1 = RNA_seq
    #Dataset_2 = NCI_60_RNA
    #Dataset_1 = CCLE_proteins
    #Dataset_2 = NCI_60_proteins
    #Dataset_1 = CCLE_Met_Common 
    #Dataset_2 = NCI_60_metabolites_common
    Comparison_type <- paste(deparse(substitute(Dataset_1)),deparse(substitute(Dataset_2)))
    Dataset1 <- as.data.frame(Dataset_1)
    Dataset2 <- as.data.frame(Dataset_2)
    Common_Cell_lines <- intersect(colnames(Dataset1),colnames(Dataset2)) #Common Samples
    Common_proteins <- intersect(rownames(Dataset1),rownames(Dataset2)) #Common Features
    
    Master_pathways <- Master_pathways[!duplicated(Master_pathways$ID),]
    ##subsetting for specific master pathways
    # Master_pathways <- Master_pathways %>% 
    #   subset(Pathway %in% c("Amino acid","Nucleotide","TCA cycle"))
    
    Master_pathways_ordered <- Master_pathways[Master_pathways$ID %in% Common_proteins,]  %>%
      left_join(humankegg_df, by = c("ID" = "Uniprot"))%>% arrange(Pathway,SubPathway) %>%
      subset(!duplicated(ID))
      #left_join(Gene_pathways[,c("Uniprot","SubPathway")], by = c("ID" = "Uniprot")) %>% subset(!duplicated(ID))
    Master_pathways_ordered[is.na(Master_pathways_ordered)] <- "Not Known"
    
      
    
   #test that Dataset1[1,1] =! Dataset2 [1,1]
    Subset_and_matrix <- function(x, triange = c("top","bottom")){
      #x <- Dataset1
      #subsetting and ordering the Datasets
      x_sub <- x %>% .[rownames(.) %in% Common_proteins, colnames(.) %in% Common_Cell_lines]%>% 
        .[match(Common_proteins,rownames(.)),match(Common_Cell_lines,colnames(.))]
      #print(deparse(substitute(x)))
      whole_matrix <- round(cor(t(x_sub[rownames(x_sub) %in% Master_pathways_ordered$ID,])),2) %>% 
      .[match(Master_pathways_ordered$ID,row.names(.)),match(Master_pathways_ordered$ID,colnames(.))]
      cormat <- whole_matrix  %>%
        {if(triange == "top")  get_upper_tri(.)else get_lower_tri(.) }
      cormat[is.na(cormat)] <- 0  
      list(whole_matrix = whole_matrix,
           matrix = cormat,
           df = x_sub)
      #cormat
    }
    list1 <- Subset_and_matrix(Dataset_1,"top")
    list2 <- Subset_and_matrix(Dataset_2, "bottom")
    Dataset1_sub <- list1$df %>% as.data.frame()
    Dataset2_sub <- list2$df %>% as.data.frame()
        #Combined_df <- rbind(cormat1,cormat2)  
    # #Combined_df$value[Combined_df$Var1 == Combined_df$Var2] <- NA
    # 
    # Feature_order <- Combined_df[!duplicated(Combined_df$Var1),] %>% 
    #   .[order(.$Pathway),"Var1"]
    # 
    # ### ComplexHeatMAP
    # test1 <- Master_pathways_ordered[order(Master_pathways_orderezd$Uniprot),]
    Metabolic_matrix <- as.data.frame(list1$matrix+list2$matrix) %>% as.matrix()
      
    
    set.seed(1587) # to set random generator seed
    cl <- colors(distinct = TRUE)
  
    
    Master_Pathway_colours <- sample(cl, length(unique(Master_pathways_ordered$Pathway))) %>% setNames(unique(Master_pathways_ordered$Pathway))
    Sub_Pathway_colours <- sample(cl, length(unique(Master_pathways_ordered$SubPathway))) %>% setNames(unique(Master_pathways_ordered$SubPathway))
    Metabolic_matrix[Metabolic_matrix ==2] <- 1
    ha = HeatmapAnnotation(Master_path = Master_pathways_ordered$Pathway,
                           Sub_path = Master_pathways_ordered$SubPathway,
                           col = list(Master_path = Master_Pathway_colours,
                                      Sub_path = Sub_Pathway_colours))
    side = rowAnnotation(Master_path = Master_pathways_ordered$Pathway,
                           Sub_path = Master_pathways_ordered$SubPathway,
                           col = list(Master_path = Master_Pathway_colours,
                                      Sub_path = Sub_Pathway_colours))
    
    hmap <- Heatmap(Metabolic_matrix, name = Comparison_type, 
            top_annotation = ha ,
            left_annotation = side,
            cluster_rows = F, cluster_columns = F, show_column_names  = F, show_row_names = F, 
            row_title = paste0("top = ", deparse(substitute(Dataset_1))), 
            column_title = paste0("bottom = ", deparse(substitute(Dataset_2))),
            column_title_side = "bottom",
            heatmap_width  = unit(20,"cm"),
            heatmap_height = unit(20,"cm"))
    
    # ### GGPLOT
    # Combined_df %>%
    #     subset(Var1 %in% Master_pathways$Uniprot & Var2 %in% Master_pathways$Uniprot) %>% na.omit() %>%
    #     ggplot(aes(factor(Var2, levels = Feature_order), factor(Var1,levels = Feature_order), fill = value))+
    #     geom_tile(color = "white")+
    #     scale_fill_gradient2(low = "blue", high = "red", mid = "white", na.value = "black",
    #                          midpoint = 0, limit = c(-1,1), space = "Lab", 
    #                          name="Pearson\nCorrelation") +
    #     theme_minimal()+ 
    #     theme(axis.text.x = element_blank(), axis.text.y = element_blank())+
    #     coord_fixed() + labs(x = deparse(substitute(Dataset_1)),deparse(substitute(Dataset_2)))#+
    # facet_grid(Pathway~., scales = "free_x")
    
    
    #Correlating corresponding Samples Correlation of only paired values which allows for NA values
    Cell_line_correlation <- map2_dbl(
        Dataset1_sub, 
        Dataset2_sub, 
        Correlation_pairwise) 
    
    Cell_line_perm <- purrr::rerun(10, 
                                   map2_dbl(
                                       Dataset1_sub[,sample(ncol(Dataset1_sub))],#Shuffle column-wise and performing sample correlation as above
                                       Dataset2_sub,
                                       Correlation_pairwise)) %>% unlist()
    
    gene_correlation <- pmap(
        list(pmap(Dataset1_sub, c, use.names = T), # unlisting feature rows into vectors into lists and correlating
             pmap(Dataset2_sub, c, use.names = T)),
        Correlation_pairwise) %>% unlist() %>% setNames(Common_proteins) 
    
    gene_correlation_perm <- purrr::rerun(10, #permutation by row shuflfing and correlation
                                          pmap(list(pmap(Dataset1_sub[sample(nrow(Dataset1_sub)),], 
                                                         c, use.names = T), #Shuffle row-wise
                                                    pmap(Dataset2_sub, c, use.names = T)),
                                               Correlation_pairwise) %>% unlist()) %>% unlist()
    
    # Imputing both datasets with Media to derive overall correlation
    Dataset1_sub_imp <- naniar::impute_median_all(Dataset1_sub)
    Dataset2_sub_imp <- naniar::impute_median_all(Dataset2_sub)
    
    Matri_Corr <- MatrixCorrelation::r1(Dataset1_sub_imp,Dataset2_sub_imp) # whole matrix correlation
    
    Matrix_perm <- purrr::rerun(10, #whole matrix permutation of Dataset 1 col-wise shuffle and Dataset 2 row-wise shuffle and correlation
                            MatrixCorrelation::r1(Dataset1_sub_imp[,sample(ncol(Dataset1_sub_imp))],
                                                  Dataset2_sub_imp[sample(nrow(Dataset2_sub_imp)),])) %>% 
        unlist() # permutation
    
    Hist_plot <- function(Data,Perm){
      #Data = Cell_line_correlation
      #Perm = Cell_line_perm
        Sample_Corr <- data.frame(Samples = c(Data,unlist(Perm)),
                              Type = rep(c("Data","Perm"), times = c(length(Data),length(unlist(Perm)))))
        p <- ggplot2::ggplot(Sample_Corr,aes(Samples)) +
    geom_histogram(aes(y=..density..,fill = Type ),      # Histogram with density instead of count on y-axis
                   binwidth=.01,alpha=0.1) +
    geom_density(alpha=.5, aes(colour = Type))+
            ggtitle(paste(Comparison_type,deparse(substitute(Data))))
    }
    
    Feature_quartiles <- rowMeans(Dataset2_sub, na.rm = T) %>% data.frame(Feature = names(.),
                                                           Mean_Abundance = .) %>%
        mutate(quartile = as.factor(ntile(Mean_Abundance, 10))) %>% left_join(stack(gene_correlation), by = c("Feature" = "ind"))
    
    Feature_expression_cor_plot <- ggplot(Feature_quartiles, aes(values,after_stat(density), colour =quartile)) +
        geom_freqpoly(binwidth = 0.1) +
        ggtitle(paste("Feature Expression",Comparison_type))
    
    Sample_lineage <-  Cell_line_correlation %>% data.frame(Sample = names(.), Correlation = .) %>%
      left_join(sample_info[,c("stripped_cell_line_name", "lineage")], by = c("Sample" = "stripped_cell_line_name")) %>%
      left_join(Derms, by = c("lineage" = "Cancer_Lineage" ))
    
    Sample_lineage_cor_plot <- ggplot(Sample_lineage, aes(Correlation,after_stat(density), colour = Derm)) +
        geom_freqpoly(binwidth = 0.01) +
        ggtitle(paste("Sample Lineage",Comparison_type))
    
    
    list(Sample_Corr = list(Sample_lineage = Sample_lineage, 
                            Cell_line_perm = Cell_line_perm, 
                            Histogram = Hist_plot(Cell_line_correlation,Cell_line_perm),
                            Sample_lin_cor_plot = Sample_lineage_cor_plot),
         Feature_Corr = list(Features_quart = Feature_quartiles, 
                             Gene_corr_perm = gene_correlation_perm, 
                             Histogram = Hist_plot(gene_correlation, gene_correlation_perm), 
                             Feature_cor_plot = Feature_expression_cor_plot),
         Matrix_Corr= list(Matri_Corr = Matri_Corr, 
                           Matrix_perm = Matrix_perm,
                           Histogram =   Hist_plot(Matri_Corr,Matrix_perm)),
         Heatmap= list(Dataset1_matri = list1$whole_matrix,
                       Dataset2_matri = list2$whole_matrix,
                       Plot_Heatmap = hmap))
    
}

### Metabolic Comparison ###
Metabolite_mapping <- read.csv(Metabolite_mapping_file)
NCI_60_metabolites <-  NCI_60_metabolites %>% separate_rows(Annotation.ID, sep = " ; ") %>%
  subset(str_detect(Annotation.ID, "^H|C") & !duplicated(Annotation.ID)) %>%
  mutate(Annotation.ID =  str_replace_all(Annotation.ID,"HMDB", "HMDB00")) 

Common_Metabolites <- Metabolite_mapping[(Metabolite_mapping$KEGG %in% (NCI_60_metabolites$Annotation.ID)) |
                                          (Metabolite_mapping$HMDB %in% (NCI_60_metabolites$Annotation.ID)) ,]

NCI_60_metabolites_common <- NCI_60_metabolites[str_detect(NCI_60_metabolites$Annotation.ID, 
                                                           paste(c(Common_Metabolites$HMDB,Common_Metabolites$KEGG), collapse = "|")),]

tt1=NCI_60_metabolites_common$Annotation.ID[str_detect(NCI_60_metabolites_common$Annotation.ID,"^C")]
NCI_60_metabolites_common$Annotation.ID[str_detect(NCI_60_metabolites_common$Annotation.ID,"^C")] <- Metabolite_mapping[Metabolite_mapping$KEGG %in% tt1,] %>%  .[match(tt1, .$KEGG),] %>% pull(HMDB)

NCI_60_metabolites_common <- NCI_60_metabolites_common%>% 
  subset(!duplicated(Annotation.ID)) %>%
  column_to_rownames("Annotation.ID")

CCLE_Met_Common <- Metabolites[rownames(Metabolites) %in% Common_Metabolites$Query,]


rownames(CCLE_Met_Common) <- Common_Metabolites$HMDB[match(rownames(CCLE_Met_Common),Common_Metabolites$Query)]

omics <- list(Transcriptome = Dataset_correlation(RNA_seq,NCI_60_RNA),
              Proteome = Dataset_correlation(CCLE_proteins,NCI_60_proteins),
              Metabolome = Dataset_correlation(CCLE_Met_Common,NCI_60_metabolites_common))

```

Proteomics heatmap: Purine and Pyrimidine in nucleotide = opposite trends?
AA metabolism: Proteosome and Ribosomal genes
Other pathways might have too many subpathways for pattern to emerge
TCA: Not Known pathway genes

# Sample correlation

```{r Sample_cor}

map(omics, ~.x$Sample_Corr$Histogram)

```

# Feature correlation

```{r Feature_Corr}
map(omics, ~.x$Feature_Corr$Histogram)

```
# Essentiality correlation

```{r Essentiality}
### Transcriptomic Comparison ###
Essential_vs_Not_Genes <- omics$Transcriptome$Feature_Corr$Features_quart %>%
    mutate(Essentiality = if_else(Feature %in% Essential_genes, "Essential", "Not_Essential"))

ggplot(Essential_vs_Not_Genes, aes(values,after_stat(density), colour = Essentiality)) +
        geom_freqpoly(binwidth = 0.1) +
        ggtitle(paste("mRNA essentiality Correlation"))

```


# Matrix correlation

```{r Matrix_Corr}
map(omics, ~.x$Matrix_Corr$Histogram)
```

# Sample Lineage

```{r Sample_lineage}
map(omics, ~.x$Sample_Corr$Sample_lin_cor_plot)

```


# Feature Gene expression Corr

```{r Feature_expre}
map(omics, ~.x$Feature_Corr$Feature_cor_plot)
```

1 is the bottom quartile and 4 the top quartile


# Heatmaps

```{r heatmaps, fig.width = 18, fig.height=18}
map(omics, ~.x$Heatmap$Plot_Heatmap)
```
Here we see that the proteins might actually be more biologically relevant than the mRNA and the 2 datasets are comparable. The CCLE proteome has more NAs but this is not shown here due to 'pairwise' correlation. But when the data is present, the comparison is good

# Transcript-Proteom Heatmap
```{r mix_heatmap, fig.width = 18, fig.height=18}
Genes_matrix <- omics$Transcriptome$Heatmap$Dataset1_matri 
Proteins_Matrix <- omics$Proteome$Heatmap$Dataset1_matri 

Gene_mapping <- HUMAN_9606_idmapping %>%
  subset((Type == "Gene_Name") & (Uniprot %in% rownames(Proteins_Matrix)),-2) %>%
  subset(!(duplicated(Uniprot) |  duplicated(ID))) %>%
  .[match(rownames(Proteins_Matrix),.$Uniprot),] %>% .$ID

Proteins_Matrix <- Proteins_Matrix %>% 
  magrittr::set_rownames(Gene_mapping) %>% 
  magrittr::set_colnames(Gene_mapping)

common_genes <- intersect(rownames(Genes_matrix), rownames(Proteins_Matrix))
Master_pathways <- Master_pathways[!duplicated(Master_pathways$ID),]
Master_pathways_ordered <- Master_pathways[Master_pathways$ID %in% common_genes,] %>% arrange(Pathway)# %>%
      #left_join(humankegg_df, by = c("ID" = "Uniprot")) %>% subset(!duplicated(ID))
      #left_join(Gene_pathways[,c("Uniprot","SubPathway")], by = c("ID" = "Uniprot")) %>% subset(!duplicated(ID))
Master_pathways_ordered[is.na(Master_pathways_ordered)] <- "Not Known"
common_genes <- common_genes[match(Master_pathways_ordered$ID, common_genes)]

Genes_matrix <- Genes_matrix %>% 
  .[rownames(.) %in% common_genes,colnames(.) %in% common_genes] %>%
  .[match(common_genes,rownames(.)),match(common_genes,colnames(.))] %>% 
  get_upper_tri()
Genes_matrix[is.na(Genes_matrix)] <- 0

Proteins_Matrix <- Proteins_Matrix %>% 
  .[rownames(.) %in% common_genes,colnames(.) %in% common_genes] %>%
  .[match(common_genes,rownames(.)),match(common_genes,colnames(.))] %>%
  get_lower_tri()
Proteins_Matrix[is.na(Proteins_Matrix)] <- 0



Combined_matrix <- as.data.frame(Genes_matrix+Proteins_Matrix) %>% as.matrix()
      
    
set.seed(1587) # to set random generator seed
cl <- colors(distinct = TRUE)
  
Master_Pathway_colours <- sample(cl, length(unique(Master_pathways_ordered$Pathway))) %>% setNames(unique(Master_pathways_ordered$Pathway))
#Sub_Pathway_colours <- sample(cl, length(unique(Master_pathways_ordered$SubPathway))) %>% setNames(unique(Master_pathways_ordered$SubPathway))
Combined_matrix[Combined_matrix ==2] <- 1
ha = HeatmapAnnotation(Master_path = Master_pathways_ordered$Pathway,
                           #Sub_path = Master_pathways_ordered$SubPathway,
                           col = list(Master_path = Master_Pathway_colours))#,
                                      #Sub_path = Sub_Pathway_colours))
side = rowAnnotation(Master_path = Master_pathways_ordered$Pathway,
                           #Sub_path = Master_pathways_ordered$SubPathway,
                           col = list(Master_path = Master_Pathway_colours))#,
                                      #Sub_path = Sub_Pathway_colours))
    
hmap <- Heatmap(Combined_matrix, #name = Comparison_type, 
            top_annotation = ha ,
            left_annotation = side,
            cluster_rows = F, cluster_columns = F, show_column_names  = F, show_row_names = F, 
            row_title = paste0("top = ", "Transcriptome"), 
            column_title = paste0("bottom = ", "Proteome"),
            column_title_side = "bottom")
hmap

```

```{r Gene_protein_rank}
#this is done with the overlapping terms, should be done with the individual terms
RNA_feature_ranking <- omics$Transcriptome$Feature_Corr$Features_quart %>% 
  select(-values) %>% setNames(c("Gene_name","Mean_RNA_Abundance","Gene_Rank"))
Protein_feature_ranking <- omics$Proteome$Feature_Corr$Features_quart %>% 
  select(-values) %>% setNames(c("Uniprot","Mean_Prot_Abundance","Protein_Rank"))


ensembl <- biomaRt::useMart('ensembl', dataset="hsapiens_gene_ensembl")
annotation <- biomaRt::getBM(attributes=c("uniprotswissprot", "uniprot_gn_symbol"), 
                             filters="uniprotswissprot", 
                             values=Protein_feature_ranking$Uniprot, mart=ensembl)

annotation <- subset(HUMAN_9606_idmapping, (Type == "Gene_Name") & 
                       (Uniprot %in% Protein_feature_ranking$Uniprot)) 

Rankings <- RNA_feature_ranking %>%
  full_join(annotation, by = c("Gene_name" = "ID")) %>%
  full_join(Protein_feature_ranking, by = "Uniprot")

hist(as.numeric(Rankings$Mean_RNA_Abundance[!is.na(Rankings$Protein_Rank)]))

breaks <- seq(min(Rankings$Mean_RNA_Abundance, na.rm = T)-0.1,
              max(Rankings$Mean_RNA_Abundance, na.rm = T)+0.1,
                  by = 0.2)
with(Rankings, hist(Rankings$Mean_RNA_Abundance, col = rgb(1,0,0,0.5), breaks= breaks))
with(Rankings, hist(Rankings$Mean_RNA_Abundance[!is.na(Rankings$Protein_Rank)], col = rgb(0,0,1,0.5), add = T, breaks= breaks))

```

```{r Information_flow}
IntraFeatureCorrelation <- function(dataset1,dataset2){
  #create tests lineage type and cell lines?
  
  #dataset1 <- NCI_60_proteins
  #dataset2 <- NCI_60_RNA
  # dataset1 <- CCLE_proteins
  # dataset2 <- RNA_seq
  Comparison_type <- paste(deparse(substitute(dataset1)),deparse(substitute(dataset2)))
  Common_Cell_lines <- intersect(colnames(dataset1),
                                 colnames(dataset2))
  
  dataset1 <- as.data.frame(dataset1) %>% .[colnames(.) %in% Common_Cell_lines]%>% 
        .[,match(Common_Cell_lines,colnames(.))]
  dataset2 <- as.data.frame(dataset2) %>% .[colnames(.) %in% Common_Cell_lines]%>% 
        .[,match(Common_Cell_lines,colnames(.))]
  
  sample_info_dataset <- sample_info[sample_info$stripped_cell_line_name %in% Common_Cell_lines,]
  
  lineage_types <- sample_info_dataset$lineage %>% table() %>% .[. >4] %>% names()
  # Cell_line <- sample_info_dataset$stripped_cell_line_name[sample_info_dataset$lineage %in% lineage_types] %>% 
  #   .[match(Common_Cell_lines,Cell_line)]

list1 <-     map(lineage_types, ~dplyr::select(dataset1,any_of(sample_info_dataset$stripped_cell_line_name[sample_info_dataset$lineage == .x]))) %>%     set_names(lineage_types)


   list2 <- map(lineage_types, ~dplyr::select(dataset2,any_of(sample_info_dataset$stripped_cell_line_name[sample_info_dataset$lineage == .x]))) %>%     set_names(lineage_types) 
  #dataset1[is.na(dataset1)] <- 0
  #dataset2[is.na(dataset2)] <- 0
  annotation <- subset(HUMAN_9606_idmapping, (Type == "Gene_Name") & 
                       (Uniprot %in% rownames(dataset1) & 
                          (ID %in% rownames(dataset2))))
  Matching <- rownames(dataset1)[rownames(dataset1) %in% annotation$Uniprot]
  
  # 
  # y = list1[["breast"]]
  # z = list2[["breast"]]
  # x = "Q99453"
  Correlation_intrafeature <- function(x,y,z){
    Correlation_pairwise(y[rownames(y) == x,] %>% unlist(),
                         z[rownames(z) == annotation$ID[annotation$Uniprot == x],] %>% unlist())
  }
  Safe_intraFeature <- safely(Correlation_intrafeature)
  
  Safe_intraFeature_tissue <- function(a,b,c){
    map(Matching, ~Safe_intraFeature(.x,a, b)) %>% 
    set_names(Matching) %>% map("result") %>%
    compact() %>% unlist() %>% stack() %>% 
    setNames(c("Corr","Uniprot")) %>% 
    left_join(Master_pathways, by = c("Uniprot" = "ID")) %>%
      mutate(Tissue = c)
  }
  
  Tissue_Intracorrelation <- pmap_df(list(list1,list2,lineage_types),Safe_intraFeature_tissue)
  Tissue_Intracorrelation$Pathway[is.na(Tissue_Intracorrelation$Pathway)] <- "Not Known"
  
  Pan_Tissue_Intracorrelation <- map(Matching, ~Safe_intraFeature(.x,dataset1, dataset2)) %>% 
    set_names(Matching) %>% map("result") %>%
    compact() %>% unlist() %>% stack() %>% 
    setNames(c("Corr","Uniprot")) %>% 
    left_join(Master_pathways, by = c("Uniprot" = "ID")) %>% mutate(Tissue = "all")
  
  PanPathway_cor <- ggplot(Pan_Tissue_Intracorrelation,aes(Corr,fill = Pathway, colour = Pathway)) +
         geom_density(alpha = 0.05) +
        ggtitle(paste("IntraFeature Corr by Pathway",Comparison_type))
  
  
  
  Pathway_per_tissue <- map(unique(Tissue_Intracorrelation$Pathway), 
                                 ~ggplot(Tissue_Intracorrelation[Tissue_Intracorrelation$Pathway ==.x,],
                                         aes(Corr,fill = Tissue, colour = Tissue)) +
                                   geom_density(alpha = 0.05) +
                                   ggtitle(paste("IntraFeature Corr by Pathway",.x))) %>% 
    set_names(unique(Tissue_Intracorrelation$Pathway))
  
  Pathway_per_tissue_grid <- ggplot(Tissue_Intracorrelation,
                                         aes(Corr,fill = Pathway,colour = Pathway))+
    geom_density(alpha = 0.1) +
    facet_wrap(~ Tissue)+ 
    ggtitle(paste("IntraFeature Corr by Pathway",""))
  
  Tissue_per_pathway<- map(unique(Tissue_Intracorrelation$Tissue), 
                                 ~ggplot(Tissue_Intracorrelation[Tissue_Intracorrelation$Tissue ==.x,],
                                         aes(Corr,fill = Pathway, colour = Pathway)) +
                                   geom_density(alpha = 0.05) +
                                   ggtitle(paste("IntraFeature Corr by Pathway",.x))) %>% 
    set_names(unique(Tissue_Intracorrelation$Tissue))

  Tissue_specific_pathway_grid <- ggplot(Tissue_Intracorrelation,
                                         aes(Corr,fill = Tissue,colour = Tissue))+
    geom_density(alpha = 0.1) +
    facet_wrap(~ Pathway)+ 
    ggtitle(paste("IntraFeature Corr by Pathway",""))
  
    PanCancer_Corr <- ggplot(Tissue_Intracorrelation, aes(Corr,fill = Tissue, colour = Tissue)) +
                                   geom_density(alpha = 0.05) +
                                   ggtitle("IntraFeature Corr Across Tissue")
  
  list(Pan_Tissue_Intracorrelation_df = Pan_Tissue_Intracorrelation,
       PanCancer_Correlation_plot = PanCancer_Corr,
       Tissue_specific_pathway_grid = Tissue_specific_pathway_grid,
       Pathway_per_tissue_grid = Pathway_per_tissue_grid,
       Tissue_per_pathway_plot = Tissue_per_pathway,
       Pathway_per_tissue_plot = Pathway_per_tissue,
       PanPathway_cor_plot = PanPathway_cor,
       Tissue_Intracorrelation_df = Tissue_Intracorrelation)
  

}



IntraFeature_CCLE <- IntraFeatureCorrelation(CCLE_proteins,
                                             RNA_seq)  

IntraFeature_NCI <- IntraFeatureCorrelation(NCI_60_proteins,
                                             NCI_60_RNA)  
IntraFeature_CCLE[str_detect(names(IntraFeature_CCLE),"plot|grid")] %>% imap(print)
IntraFeature_NCI[str_detect(names(IntraFeature_NCI),"plot|grid")] %>% imap(print)
```

```{r Pathway_specific_trends}
map(unique(IntraFeature_NCI[["Tissue_Intracorrelation_df"]][,"Pathway"]),
    ~IntraFeature_NCI$Tissue_Intracorrelation_df %>% 
  mutate(Colour = if_else(.[,"Pathway"] == .x,"blue","gray")) %>% 
  ggplot(aes(Corr,fill = Colour,colour = Colour))+
    geom_density(alpha = 0.1) +
  scale_color_identity(labels = c(blue = .x,gray = "Others"),guide = "legend")+
    scale_fill_identity(labels = c(blue = .x,gray = "Others"))+
    facet_wrap(~ eval(parse(text = "Tissue")))+ 
    ggtitle(paste("IntraFeature Corr by Pathway","")))

map(unique(IntraFeature_CCLE[["Tissue_Intracorrelation_df"]][,"Pathway"]),
    ~IntraFeature_CCLE$Tissue_Intracorrelation_df %>% 
  mutate(Colour = if_else(.[,"Pathway"] == .x,"blue","gray")) %>% 
  ggplot(aes(Corr,fill = Colour,colour = Colour))+
      scale_color_identity(labels = c(blue = .x,gray = "Others"),guide = "legend")+
    scale_fill_identity(labels = c(blue = .x,gray = "Others"))+
    geom_density(alpha = 0.1) +
    facet_wrap(~ Tissue)+ 
    ggtitle(paste("IntraFeature Corr by Pathway","")))


map(unique(IntraFeature_CCLE$Tissue_Intracorrelation_df$Tissue),
~IntraFeature_CCLE$Tissue_Intracorrelation_df %>% 
  mutate(Colour = if_else(.$Tissue == .x,"blue","gray")) %>%
  ggplot(aes(Corr,fill = Colour,colour = Colour))+
    scale_color_identity(labels = c(blue = .x,gray = "Others"),guide = "legend")+
    scale_fill_identity(labels = c(blue = .x,gray = "Others"))+
    geom_density(alpha = 0.1) +
    facet_wrap(~ Pathway)+ 
    ggtitle(paste("IntraFeature Corr by Pathway","")))


map(unique(IntraFeature_CCLE$Tissue_Intracorrelation_df$Tissue),
~IntraFeature_CCLE$Tissue_Intracorrelation_df %>% 
  mutate(Colour = if_else(.$Tissue == .x,"blue","gray")) %>%
  ggplot(aes(Corr,fill = Colour,colour = Colour))+
    scale_color_identity(labels = c(blue = .x,gray = "Others"),guide = "legend")+
    scale_fill_identity(labels = c(blue = .x,gray = "Others"))+
    geom_density(alpha = 0.1) +
    facet_wrap(~ Pathway)+ 
    ggtitle(paste("IntraFeature Corr by Pathway","")))
```

```{r eIF_diff}

PivotDasetTissue <- function(Dataset){ # this is for proteomics, adapt to RNAseq
  #Dataset = CCLE_proteins
    sample_info_dataset <- sample_info %>% .[.$stripped_cell_line_name %in% colnames(Dataset),]
  lineage_types <- sample_info_dataset$lineage %>% table() %>% .[. >4] %>% names()
  list1 <- map(lineage_types,
               ~select(as.data.frame(Dataset),
                       any_of(sample_info_dataset$stripped_cell_line_name[sample_info_dataset$lineage == .x]))) %>%
    set_names(lineage_types)
  map_dfr(list1, ~tibble::rownames_to_column(.x,var = "Uniprot")) %>%
    pivot_longer(-Uniprot, names_to = "Cell_line", values_to = "Abundance")
}
CCLE_pro_pivotted <- PivotDasetTissue(CCLE_proteins)
NCI_pro_pivotted <- PivotDasetTissue(NCI_60_proteins)




Plot_markers_Abundance <- function(Dataset, markers){
  #Dataset <- NCI_pro_pivotted
  #markers <-  Uniprot_eIF
  Dataset_name <- deparse(substitute(Dataset))
  markers <- markers[markers %in% Dataset$Uniprot]
  Dataset <- Dataset %>% filter(Uniprot %in% markers) %>% 
      left_join(sample_info_dataset[,c("stripped_cell_line_name","lineage")],
                by = c("Cell_line" = "stripped_cell_line_name"))
  map(markers, ~Dataset %>% filter(Uniprot == .x) %>% group_by(lineage) %>%
        ggplot(aes(x = lineage, colour = lineage,
                   y = Abundance)) +
    geom_boxplot() +
      geom_point() +
      ggtitle(paste(.x,HUMAN_9606_idmapping %>% filter(Uniprot == .x &
                                                         Type == "UniProtKB-ID") %>%
                      pull(ID), Dataset_name,"Abundance")))
}
Hypox_Prolif <- HUMAN_9606_idmapping %>% filter((str_detect(ID, 'HIF') & Type == "UniProtKB-ID")| (ID == "KI67_HUMAN")) %>%
  pull(Uniprot) %>% unique()
Uniprot_eIF <- read_tsv(here::here("Project_Datasets","uniprot_eukaryotic_translation_factor.tab")) %>% pull(Entry)


eIFs_CCLE <- Plot_markers_Abundance(CCLE_pro_pivotted,Uniprot_eIF)
eIFs_NCI <- Plot_markers_Abundance(NCI_pro_pivotted,Uniprot_eIF)
Hypoxia_prolif_CCLE <- Plot_markers_Abundance(CCLE_pro_pivotted,Hypox_Prolif)
Hypoxia_prolif_NCI <- Plot_markers_Abundance(NCI_pro_pivotted,Hypox_Prolif)

save(file = here("Project_Datasets","eIF_CCLE_NCI_hypoxia.RData"),eIFs_CCLE,eIFs_NCI,Hypoxia_prolif_CCLE,Hypoxia_prolif_NCI)
load(here("Project_Datasets","eIF_CCLE_NCI_hypoxia.RData"))
map(list(eIFs_CCLE,eIFs_NCI,Hypoxia_prolif_CCLE,Hypoxia_prolif_NCI), print)

```

```{r marker_gene_corr}
# TranslationalEfficiency <- function(Proteome = dataset1,
#                                     Transcriptome = dataset2){
  #create tests lineage type and cell lines?

  #dataset1 <- NCI_60_proteins
  #dataset2 <- NCI_60_RNA
  dataset1 <- CCLE_proteins
  dataset2 <- RNA_seq

  dataset1 <- CCLE_proteins %>% apply(1, scales::rescale) %>% t()
  dataset2 <- dataset2
  dataset1[dataset1 == 0 ] <- 0.000001
  dataset2[dataset2 == 0 ] <- 0.000001
  Comparison_type <- paste(deparse(substitute(dataset1)),deparse(substitute(dataset2)))
  Common_Cell_lines <- intersect(colnames(dataset1),
                                 colnames(dataset2))

  dataset1 <- as.data.frame(dataset1) %>% .[colnames(.) %in% Common_Cell_lines]%>%
        .[,match(Common_Cell_lines,colnames(.))]
  dataset2 <- as.data.frame(dataset2) %>% .[colnames(.) %in% Common_Cell_lines]%>%
        .[,match(Common_Cell_lines,colnames(.))]

  sample_info_dataset <- sample_info[sample_info$stripped_cell_line_name %in% Common_Cell_lines,]

  lineage_types <- sample_info_dataset$lineage %>% table() %>% .[. >4] %>% names()
  # Cell_line <- sample_info_dataset$stripped_cell_line_name[sample_info_dataset$lineage %in% lineage_types] %>%
  #   .[match(Common_Cell_lines,Cell_line)]
  list1 <- map(lineage_types, ~dplyr::select(dataset1,any_of(sample_info_dataset$stripped_cell_line_name[sample_info_dataset$lineage == .x]))) %>%
    set_names(lineage_types)
  list2 <- map(lineage_types, ~dplyr::select(dataset2,any_of(sample_info_dataset$stripped_cell_line_name[sample_info_dataset$lineage == .x]))) %>%
    set_names(lineage_types)



  #dataset1[is.na(dataset1)] <- 0
  #dataset2[is.na(dataset2)] <- 0
  annotation <- subset(HUMAN_9606_idmapping, (Type == "Gene_Name") &
                       (Uniprot %in% rownames(dataset1) &
                          (ID %in% rownames(dataset2))))
  Matching <- rownames(dataset1)[rownames(dataset1) %in% annotation$Uniprot]

  #
  Translation_Ratio <- function(x,y,z){
    tst1 <-  (y[rownames(y) == x,] %>% unlist())/ (z[rownames(z) == annotation$ID[annotation$Uniprot == x],] %>% unlist())
    tst1 %>% .[!is.na(.) & !is.infinite(.)] %>% stack %>% setNames(c("TE","Cell_line")) %>% mutate(Uniprot = x)
  }
  Safe_Translation_Ratio <- safely(Translation_Ratio)

  Safe_Translation_Ratio_tissue <- function(a,b){
    # a = list1$breast
    # b = list2$breast
    # .x = Matching[2]
    #
    map(Matching, ~Safe_Translation_Ratio(.x,a, b)) %>%
     map("result") %>% compact() %>% purrr::reduce(rbind) %>%
    left_join(Master_pathways, by = c("Uniprot" = "ID"))
  }
  library(future)
  future::plan(multisession, workers = 8)
  Tissue_Translation_Ratio_v <- future::future({map2_df(list1,list2,Safe_Translation_Ratio_tissue)},seed = T)
  Tissue_Translation_Ratio <- future::value(Tissue_Translation_Ratio_v)
  future::plan(sequential)

  
  Tissue_Translation_Ratio <- map2_df(list1["ovary"],list2["ovary"],Safe_Translation_Ratio_tissue)
  Tissue_Translation_Ratio$Pathway[is.na(Tissue_Translation_Ratio$Pathway)] <- "Not Known"

  Tissue_Translation_Ratio <- Tissue_Translation_Ratio %>%
    mutate(Cell_line = as.character(Cell_line)) %>%
             group_split(Uniprot)
  Tissue_Translation_Ratio <- map(Tissue_Translation_Ratio, "Uniprot") %>% unlist() %>%unique() %>%
    setNames(Tissue_Translation_Ratio,.)

  # x <- Combinations %>% pull(Proteins)
  # y <- Combinations %>% pull(eIFs)

  eIFCorrProtein <- function(x,y){
    #x = map(Combinations$Proteins[1:2], ~Tissue_Translation_Ratio[[.x]])[[1]]
    #y = map(Combinations$eIFs[1:2], ~dataset1_testing[[.x]])[[1]]
    common_lines <- intersect(colnames(y), x %>% dplyr::select_if(~ !any(is.na(.))) %>% pull(Cell_line))

    correlation <- cor(x %>% subset(Cell_line %in% common_lines) %>% .[match(common_lines,.$Cell_line),] %>% pull(TE),
                       y %>% dplyr::select(all_of(common_lines)) %>% .[,match(common_lines,colnames(.))] %>% unlist())

    data.frame(Protein = unique(x$Uniprot), Factor = unique(y$Uniprot), Correlation = correlation)
    #browser()
  }

Proteins <- names(Tissue_Translation_Ratio)
  eIFs <- dataset1[row.names(dataset1) %in% Uniprot_eIF,] %>% row.names()
  Combinations <- tidyr::crossing(Proteins, eIFs) #%>%
    #head(5)
dataset1_testing <-  dataset1 %>% 
  rownames_to_column(var = "Uniprot") %>%
  subset(Uniprot %in% Uniprot_eIF) %>%
  group_split(Uniprot) %>% set_names(.,map_chr(.x = .,"Uniprot"))# 
options(future.globals.maxSize= 891289600)
  eIFCorrProtein_safe <- safely(eIFCorrProtein)
  library(future)
  plan(multisession, workers = 8)
  eIF_correlations_f <-  future({map2(map(Combinations$Proteins, ~Tissue_Translation_Ratio[[.x]]),
                                      map(Combinations$eIFs, ~dataset1_testing[[.x]]),eIFCorrProtein_safe)},
           seed = T)
  eIF_correlations <- value(eIF_correlations_f) %>%
  map("result") %>%  compact()%>% 
  purrr::reduce(rbind)
  plan(sequential)
  

 #save(file = here("Project_Datasets","eIF_CCLE_Correlations.RData"),eIF_correlations)
 save(file = here("Project_Datasets","eIF_CCLE_Correlations_ovary.RData"),eIF_correlations)
```
```{r gganimate}
importing_EIF_corr <- function(x){
  load(here("Project_Datasets",x))
  eIF_correlations
}
files <- list.files(here("Project_Datasets"), pattern = "eIF_CCLE_Correlations_")
eIFs_corr <- purrr::map(files,importing_EIF_corr) %>% setNames(str_match(files,"Correlations_([:graph:]*?).RData")[,2])
  
```

```{r PCA, fig.width = 10, fig.height=10}
library(biomaRt)
ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")

gganimate_pca <- function(x,y){
  eIF_correlations <- x 
eIF_correlations <- eIF_correlations %>% 
  left_join(Master_pathways %>% 
              dplyr::select(!c("Type")), by = c("Protein" = "ID"))


eIF_correlations[is.na(eIF_correlations)] <- "Not Known"
eIF_correlations$Correlation <- eIF_correlations$Correlation %>% as.numeric()
# Per_eIF_corr <-  map(unique(eIF_correlations$Factor), 
#                                  ~ggplot(eIF_correlations[eIF_correlations$Factor ==.x,],
#                                          aes(Correlation,fill = Pathway, colour = Pathway)) +
#                                    geom_density(alpha = 0.05) +
#                                    ggtitle(paste("EEF Corr by Pathway",.x))) %>% 
#     set_names(unique(eIF_correlations$Factor))
# 
# walk(Per_eIF_corr, plot)
#devtools::install_github("tidymodels/broom")
#library(broom)  
library(cowplot)

eIF_correlations_wide <- eIF_correlations %>%
  left_join(HUMAN_9606_idmapping %>%
              subset(Type == "Gene_Name") %>%
              dplyr::select(-Type), by = c("Protein" = "Uniprot")) %>%
  mutate(across(where(is.factor), as.character),
         Protein = paste(ID,Protein, sep = "_")) %>%
  subset(!duplicated(dplyr::select(.,Factor,Protein))) %>%
  pivot_wider(id_cols = Factor, names_from = Protein, values_from = Correlation) 


#                                                         
# eIF_correlations_wide <- eIF_correlations_wide %>%
#   inner_join(Translation_factors, by = c("ID" = "external_gene_name"))

#eIF_correlations_wide_metabolism <-  eIF_correlations_wide #%>%
  #dplyr::select(any_of(c("Factor","ID",Master_pathways$ID)))
eIF_correlations_wide <- eIF_correlations_wide[,1:100] %>% 
  left_join(HUMAN_9606_idmapping %>%
              subset(Type == "Gene_Name") %>%
              dplyr::select(-Type), by = c("Factor" = "Uniprot")) %>%
  subset(!duplicated(.["Factor"]))%>%
  column_to_rownames("Factor") %>%
   subset(is.na(.) %>% rowSums<(ncol(.)/2)) %>%
  select_if(~!any(is.na(.)))  

pca_fit <-eIF_correlations_wide %>% 
   dplyr::select(where(is.numeric))%>%
  as.matrix()%>%
  impute::impute.knn() %>% .[["data"]]%>%
  #tidyr::drop_na()%>%
  #mutate(across(where(is.list),as.numeric)) %>%
  prcomp() #%>% # retain only numeric columns
  #prcomp(scale = TRUE)

# p <-  eIF_correlations_wide %>% 
#   column_to_rownames("Factor") %>%
#   select_if(~ !any(is.na(.))) %>%
#   #mutate(across(where(is.list),as.numeric)) %>%
#   dplyr::select(where(is.numeric))%>% as.matrix() %>% t() %>%
#    PCAtools::pca(removeVar = 0.1)
# screeplot(p, axisLabSize = 18, titleLabSize = 22)
# biplot(p, showLoadings = TRUE, lab = NULL)
# 
# plotloadings(p, components = 1:2,
#              rangeRetain = 0.05,
#              labSize = 4.0,
#              title = 'Loadings plot',
#              subtitle = 'PC1',
#              caption = 'Top 1% variables',
#              shape = 24,
#              col = c('blue', 'black', 'red'),
#              drawConnectors = TRUE)

pca_fit$x[,1:2] %>% 
  as.data.frame(.,row.names = rownames(.))%>%
  rownames_to_column("Factor")%>%
  mutate(tissue = y) %>%
  left_join(HUMAN_9606_idmapping %>% 
              subset(Type == "Gene_Name") %>%
              dplyr::select(-Type),
            by = c("Factor" = "Uniprot")) %>%
  subset(.,!duplicated(.$Factor))
  #broom::augment(eIF_correlations_wide) %>% # add original dataset back in

}
IDs <- eIF_PCA %>% pull(ID) %>% unique()
Translation_factors <- getBM(attributes=c("superfamily",'external_gene_name'),
                  filters = 'external_gene_name',
                  values = IDs, mart =ensembl) %>%
  subset(str_detect(.$superfamily,"SS")) %>% 
  subset(!duplicated(.$external_gene_name)) 
initial_positions <- eIF_PCA %>%
  subset(tissue == "blood") %>% mutate(
  initial_position = case_when(
    PC1 > 1 ~ "red",
    PC1 < -1 ~ "blue",
     TRUE ~ "gray")
) %>% dplyr::select("Factor","initial_position")

eIF_PCA <- imap_dfr(eIFs_corr,gganimate_pca)
p <- eIF_PCA %>% 
  add_count(ID) %>% 
  subset(.,n == length(unique(.$tissue))) %>%
  left_join(Translation_factors,by = c("ID" = "external_gene_name")) %>%  
  left_join(initial_positions)%>%
  ggplot(aes(PC1,PC2, fill = initial_position, label = ID)) + 
  #geom_point(size = 2.5) +
  geom_label(size =2)+
  #cowplot::theme_half_open(12) + cowplot::background_grid()# +
  theme(legend.position = "none") +
  transition_states(tissue,
                    transition_length = 3,
                    state_length = 4)+
  ggtitle('{closest_state}')

animate(p, fps=30, height = 10, width =10, units = "in", duration = 20, res = 150)

anim_save(here::here("Project_Output","PCA_eIFs_animated.gif"))

rotation_pca <- pca_fit %>%
  broom::tidy(matrix = "rotation") %>% 
  mutate(column = str_remove_all(column,"^*([:graph:]*)_")) %>%
  left_join(HUMAN_9606_idmapping %>% 
              subset(Type == "KEGG") %>%
              dplyr::select(-Type),by = c("column" = "Uniprot")) %>%
  mutate(ID = str_remove_all(ID, "hsa:"))

pca_fit %>%
  broom::tidy(matrix = "eigenvalues") %>%
  ggplot(aes(PC, percent)) +
  geom_col(fill = "#56B4E9", alpha = 0.8) +
  scale_x_continuous(breaks = 1:9) +
  scale_y_continuous(
    labels = scales::percent_format(),
    expand = expansion(mult = c(0, 0.01))
  ) +
  cowplot::theme_minimal_hgrid(12)

# define arrow style for plotting
arrow_style <- arrow(
  angle = 20, ends = "first", type = "closed", length = grid::unit(8, "pt")
)

# plot rotation matrix
pca_fit %>%
  broom::tidy(matrix = "rotation") %>%
  pivot_wider(names_from = "PC", names_prefix = "PC", values_from = "value") %>%
  arrange(PC1) %>%
  dplyr::slice(c(1:20,(nrow(.)-20):nrow(.))) %>%
  left_join(HUMAN_9606_idmapping %>% 
              subset(Type == "Gene_Name") %>%
              dplyr::select(-Type),by = c("column" = "Uniprot")) %>%
  ggplot(aes(PC1, PC2)) +
  geom_segment(xend = 0, yend = 0, arrow = arrow_style) +
  ggrepel::geom_text_repel(
    aes(label = ID), colour = "blue"
  ) +
  coord_fixed() + # fix aspect ratio to 1:1
  theme_minimal_grid(12)

gene_list_PC_1 <- pull(distinct(rotation_pca %>% filter(PC == 1) %>% arrange(desc(value))), "value", "ID")
gene_list_PC_2 <- pull(distinct(rotation_pca %>% filter(PC == 2) %>% arrange(desc(value))), "value", "ID")
de_up <- names(gene_list_PC_1)[gene_list_PC_1 > 0.02]
de_dn <- names(gene_list_PC_1)[gene_list_PC_1 < -0.02]
kk1 <- gseKEGG(geneList     = gene_list_PC_1,
               organism     = 'hsa',
               nPerm        = 1000,
               minGSSize    = 60,
               pvalueCutoff = 0.05,
               verbose      = FALSE)
enrichplot::upsetplot(kk1)
kk2 <- gseKEGG(geneList     = gene_list_PC_2,
               organism     = 'hsa',
               nPerm        = 1000,
               minGSSize    = 60,
               pvalueCutoff = 0.05,
               verbose      = FALSE)
enrichplot::upsetplot(kk2)
gene.df <- bitr(de_up, fromType = "ENTREZID",
        toType = c("ENSEMBL", "SYMBOL"),
        OrgDb = org.Hs.eg.db) 
write_tsv(gene.df, "test_genesPathAX.tsv", quote = "none")

Reactome_dot_plot <- function(list_x){
  #PC_gene_list <- gene_list_PC_1
  #threshold = 0.02
  #Direction <- "up"
  #list_x <- Reactome_dot_plot_list[[1]]
  PC_gene_list <- list_x[[1]]
  threshold <- list_x[[2]]
  Direction <- list_x[[3]]
  if(Direction == "up"){
      gene=unique(names(PC_gene_list)[PC_gene_list > threshold])
  }else{
       gene=unique(names(PC_gene_list)[PC_gene_list < -abs(threshold)])
  }
  list(enrichplot::dotplot(ReactomePA::enrichPathway(gene, 
                                                pvalueCutoff = 0.05, 
                                                readable=TRUE), 
                      showCategory = 15) + 
    ggtitle(Direction),
    enrichplot::cnetplot(setReadable(ReactomePA::enrichPathway(gene, 
                                                pvalueCutoff = 0.05, 
                                                readable=TRUE), 'org.Hs.eg.db', 'ENTREZID'), categorySize="pvalue", foldChange=PC_gene_list))
  
}
Reactome_dot_plot_list <- list(PCs = list(gene_list_PC_1,gene_list_PC_2),
     threshold = 0.02,
     Direction = c("up","down")) %>% 
  cross
map(Reactome_dot_plot_list,Reactome_dot_plot)


y <- gsePathway(geneList, 
                pvalueCutoff = 0.2,
                pAdjustMethod = "BH", 
                verbose = FALSE)
head(y)
edo2 <- DOSE::gseDO(geneList)
edo2_PC <- DOSE::gseDO(geneList)gene_list_PC
dotplot(x) + ggtitle("dotplot for ORA")
dotplot(edo2, showCategory=30) + ggtitle("dotplot for GSEA")
viewPathway("mRNA Splicing", 
            readable = TRUE, 
            foldChange = geneList)

+
```

